pkg/vere/db/lmdb.c

```diff
@@ -342,6 +342,50 @@ u3_lmdb_read(MDB_env* env_u,
   }
 }
 
+/* u3_lmdb_read_one(): load [eve_d] into [val_u].
+*/
+c3_o
+u3_lmdb_read_one(MDB_env* env_u,
+                 MDB_val* val_u,
+                 c3_d     eve_d)
+{
+  MDB_txn* txn_u;
+  MDB_dbi  mdb_u;
+  c3_w     ret_w;
+
+  //  create a read transaction
+  //
+  if ( (ret_w = mdb_txn_begin(env_u, 0, MDB_RDONLY, &txn_u)) ) {
+    mdb_logerror(stderr, ret_w, "lmdb: read one: txn_begin fail");
+    return c3n;
+  }
+
+  //  open the database in the transaction
+  //
+  if ( (ret_w =  mdb_dbi_open(txn_u, "EVENTS", 0, &mdb_u)) ) {
+    mdb_logerror(stderr, ret_w, "lmdb: read one: dbi_open fail");
+    mdb_txn_abort(txn_u);
+    return c3n;
+  }
+
+  //  set the key to [eve_d]
+  {
+    MDB_val key_u = { .mv_size = sizeof(c3_d), .mv_data = &eve_d };
+
+    if ( (ret_w = mdb_get(txn_u, mdb_u, &key_u, val_u)) ) {
+      mdb_logerror(stderr, ret_w, "lmdb: read one failed");
+      mdb_txn_abort(txn_u);
+      return c3n;
+    }
+    else {
+      //  read-only transactions are aborted when complete
+      //
+      mdb_txn_abort(txn_u);
+      return c3y;
+    }
+  }
+}
+
```

Comment:
Well, I had originally advised @matthew-levan to just use a callback, but it did strike me as a bit odd that our lmdb wrapper has a synchronous write interface and an asynchronous read interface.  For reading a whole sequence of values, it makes sense to supply a callback that gets fired on each value, but it seems useful to me to have a synchronous API in both reading and writing -- or only async APIs, but I was surprised to find both.

Having a synchronous API here allows client code to be much more straightforward -- it doesn't need a heap-allocated struct to track intermediate products.  Maybe it's more lines of code total, but I'd argue it's less mental overhead.

A counterargument is that future database APIs, such as anything that involves a distributed database, are likely to be async-only.  Then any code built on synchronous reads would have to be significantly reorganized later -- of course, that's also true for any code we have now that uses the synchronous write API, but that doesn't mean we have to make the problem worse.

Reply:
The reason I did it was because it allowed me to read the data I needed for `chop` much more quickly in terms of mental effort. 

I started working with the existing asynchronous functions and wasn't able to code a solution as quickly as I would've liked, so I trashed the approach and wrote these sync functions. 

LMDB's API is synchronous and quite straightforward, and (as a new contributor to `vere`, of course) I humbly submit that we should have similarly synchronous and straightforward wrappers to work with it.

I'm more than happy to be wrong on this point and remove the additional functions in `lmdb.c`, but it will also take some more work to implement `chop` using the original async functions.

Reply:
`u3_lmdb_read()` takes a callback, but invokes it synchronously. It's setup this way so to support reading multiple key/value pairs without committing to a particular representation or allocation strategy. The `disk.c` read API on top of this *is* async.

NB: there's an iterator-based read API that @belisarius222 and I wrote some time ago in #158.

Reply:
@matthew-levan I agree the callback-based interface to `u3_lmdb_read()` requires a little more thought to use when compared to `u3_lmdb_read_one()`, but I still think `u3_lmdb_read()` will yield a cleaner solution. To that end, can you replace `u3_lmdb_read_one()` with `u3_lmdb_read()` and see how it turns out?

Reply:
We switched to using the iterators @joemfb mentioned.  This has the advantage of not needing a callback, while also not requiring adding code to the lmdb wrapper for reading a single event, which is likely to only have this one call-site.

Reply:
`u3_lmdb_walk_*` functions are awesome.